src/internal/signalr/manager.go

```diff
@@ -0,0 +1,296 @@
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	if _, ok := m.chargers[chargerID]; ok {
+		log.Warnf("Charger '%s' is already registered", chargerID)
+
+		return
+	}
+
+	m.chargers[chargerID] = &charger{
+		handler:      handler,
+		isSubscribed: false,
+	}
+
+	m.ensureClientStarted()
+
+	if m.subscriptions != nil {
+		m.subscriptions <- chargerID
+	}
+}
+
+func (m *manager) Unregister(chargerID string) error {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	if _, ok := m.chargers[chargerID]; !ok {
+		return nil
+	}
+
+	if err := m.client.UnsubscribeCharger(chargerID); err != nil {
+		return err
+	}
+
+	delete(m.chargers, chargerID)
+
+	if len(m.chargers) == 0 {
+		if err := m.client.Close(); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (m *manager) run() {
+	states := m.client.StateC()
+	observations := m.client.ObservationC()
+
+	for {
+		select {
+		case <-m.done:
+			return
+
+		case chargerID, ok := <-m.subscriptions:
+			if !ok {
+				continue
+			}
+
+			m.handleSubscription(chargerID)
+
+		case state := <-states:
+			m.handleClientState(state)
+
+		case observation := <-observations:
+			m.handleObservation(observation)
+		}
+	}
+}
+
+func (m *manager) handleSubscription(chargerID string) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	charger, ok := m.chargers[chargerID]
+	if !ok {
+		return
+	}
+
+	if err := m.client.SubscribeCharger(chargerID); err != nil {
+		log.Warnf("Failed to subscribe charger '%s'", chargerID)
+
+		if m.subscriptions == nil {
+			return
+		}
+
+		go m.addChargerSubscription(chargerID)
+
+		return
+	}
+
+	charger.isSubscribed = true
+}
+
+func (m *manager) addChargerSubscription(chargerID string) {
+	timer := time.NewTimer(m.cfg.GetSignalRSubscribeInterval())
+	defer timer.Stop()
+
+	select {
+	case <-m.done:
+	case <-timer.C:
+		m.mu.Lock()
+		defer m.mu.Unlock()
+
+		if m.subscriptions != nil {
+			m.subscriptions <- chargerID
+		}
+	}
+}
+
+func (m *manager) handleClientState(state ClientState) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	switch state {
+	case ClientStateConnected:
+		m.subscriptions = make(chan string, 1+len(m.chargers))
+
+		for chargerID := range m.chargers {
+			select {
+			case <-m.done:
+			case m.subscriptions <- chargerID:
+			}
+		}
+
+	case ClientStateDisconnected:
+		for _, charger := range m.chargers {
+			charger.isSubscribed = false
+		}
+
+		if m.subscriptions != nil {
+			close(m.subscriptions)
+		}
+
+		m.subscriptions = nil
+
+	default:
+		log.Warnf("Unknown client state %v", state)
+	}
+}
+
+func (m *manager) handleObservation(observation Observation) {
+	if !observation.ID.Supported() {
+		return
+	}
+
+	log.Debugf("received observation: %+v", observation)
+
+	m.mu.RLock()
+	chargerHandler, ok := m.chargers[observation.ChargerID]
+	m.mu.RUnlock()
+
+	if !ok {
+		log.Warn("received observation for an unknown charger: ", observation.ChargerID)
+
+		return
+	}
+
+	if err := chargerHandler.handler.HandleObservation(observation); err != nil {
```

Comment:
if we receive an observation for a charger that's not currently subscribed, we should drop that observation.